core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+		let mut proof = Self { hashes: hashes? };
+
+		// 2. bagged peaks to the right
+		let peak_pos = family_branch
+			.last()
+			.map(|&(p, _)| p)
+			.unwrap_or(segment_last_pos);
+		if let Some(h) = pmmr.bag_the_rhs(peak_pos) {
+			proof.hashes.push(h);
+		}
+
+		// 3. peaks to the left
+		let peaks: Result<Vec<_>, _> = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x < segment_first_pos)
+			.rev()
+			.map(|p| pmmr.get_hash(p).ok_or_else(|| SegmentError::MissingHash(p)))
+			.collect();
+		proof.hashes.extend(peaks?);
+
+		Ok(proof)
+	}
+
+	/// Reconstruct PMMR root using this proof
+	pub fn reconstruct_root(
+		&self,
+		last_pos: u64,
+		segment_first_pos: u64,
+		segment_last_pos: u64,
+		segment_root: Hash,
+	) -> Result<Hash, SegmentError> {
+		let mut iter = self.hashes.iter();
+		let family_branch = pmmr::family_branch(segment_last_pos, last_pos);
+
+		// 1. siblings along the path from the subtree root to the peak
+		let mut root = segment_root;
+		for &(p, s) in &family_branch {
+			let sibling_hash = iter.next().ok_or_else(|| SegmentError::MissingHash(s))?;
+			root = if pmmr::is_left_sibling(s) {
+				(sibling_hash, root).hash_with_index(p - 1)
+			} else {
+				(root, sibling_hash).hash_with_index(p - 1)
+			};
+		}
+
+		// 2. bagged peaks to the right
+		let peak_pos = family_branch
+			.last()
+			.map(|&(p, _)| p)
+			.unwrap_or(segment_last_pos);
+
+		let rhs = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x > peak_pos)
+			.next();
+
+		if let Some(pos) = rhs {
+			root = (
+				root,
+				iter.next().ok_or_else(|| SegmentError::MissingHash(pos))?,
+			)
+				.hash_with_index(last_pos)
+		}
+
+		// 3. peaks to the left
+		let peaks = pmmr::peaks(last_pos)
+			.into_iter()
+			.filter(|&x| x < segment_first_pos)
+			.rev();
+		for pos in peaks {
+			root = (
+				iter.next().ok_or_else(|| SegmentError::MissingHash(pos))?,
+				root,
+			)
+				.hash_with_index(last_pos);
+		}
+
+		Ok(root)
+	}
+
+	/// Check validity of the proof by equating the reconstructed root with the actual root
+	pub fn validate(
+		&self,
+		last_pos: u64,
+		mmr_root: Hash,
+		segment_first_pos: u64,
+		segment_last_pos: u64,
+		segment_root: Hash,
+	) -> Result<(), SegmentError> {
+		let root =
+			self.reconstruct_root(last_pos, segment_first_pos, segment_last_pos, segment_root)?;
+		if root == mmr_root {
+			Ok(())
+		} else {
+			Err(SegmentError::Mismatch)
+		}
+	}
+}
+
+impl Readable for SegmentProof {
+	fn read<R: Reader>(reader: &mut R) -> Result<Self, Error> {
+		let n_hashes = reader.read_u64()? as usize;
+		let mut hashes = Vec::with_capacity(n_hashes);
+		for _ in 0..n_hashes {
+			let hash: Hash = Readable::read(reader)?;
+			hashes.push(hash);
+		}
+		Ok(Self { hashes })
+	}
+}
+
+impl Writeable for SegmentProof {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), Error> {
+		writer.write_u64(self.hashes.len() as u64)?;
+		for hash in &self.hashes {
+			Writeable::write(hash, writer)?;
+		}
+		Ok(())
+	}
+}
```

Comment:
```suggestion
}

#[cfg(test)]
mod tests {
	    use croaring::Bitmap;

use crate::core::hash::{Hash, DefaultHashable};
	use crate::ser::{self, Writeable, Writer};
	use super::{Segment, SegmentIdentifier, SegmentProof};

	struct TestElem(pub [u32; 4]);

	impl DefaultHashable for TestElem {}

	impl Writeable for TestElem {
		fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
			writer.write_u32(self.0[0])?;
			writer.write_u32(self.0[1])?;
			writer.write_u32(self.0[2])?;
			writer.write_u32(self.0[3])
		}
	}

	//   3
	//  / \
	// 1   2
	//
	// Two leaves at pos 1 and 2.
	// Both leaves pruned.
	// Single root remaining at pos 3.
	//
	// No leaf_data (both leaves pruned).
	// The hash provided it the parent (root) hash at pos 3.
	//
	// The root of this segment should simply be the root hash provided for pos 3.
	#[test]
	fn test_single_pruned_root() {
		let hash = Hash::default();
		let segment: Segment<TestElem> = Segment {
			identifier: SegmentIdentifier { log_size: 1, idx: 0 },
			hashes: vec![(3, hash)],
			leaf_data: vec![],
			proof: SegmentProof::empty(),
		};
		assert_eq!(hash, segment.root(3, Some(&Bitmap::create())).unwrap());
	}
}
```

Reply:
`segment.root()` does not currently handle the case where we are provided with a root hash of a pruned subtree and neither child is present in the segment.

The test above is a minimal example of this with a manually constructed segment.
I may have constructed the segment incorrectly. 
Can you confirm this is what would be provided in this situation (single hash, no leaf_data etc.)?

Reply:
This stems from the `(None, None) => None,` after popping both children off the hashes vec.
There are two scenarios to handle here - 
* both children are `None` and we are also `None` because we deep in a pruned subtree.
* both children are `None` but we are a pruned root and the hash is provided in the segment directly.

Reply:
I think this situation is already accounted for:
- Fully pruned segments cannot be created (see [here](https://github.com/mimblewimble/grin/pull/3453/files#diff-520aa054d83807bc7b454f4f7863aed94bc878ee393fa3d8df67039d47b568f8R189)).
- We do need to know the root of the segment if its sibling is not pruned, but the merkle proof of the sibling segment does contain it. That is why I still had this open point:
> I think we need a function to extract the intermediary and proof hashes for purposes of storing them in the MMR we are building

Reply:
It looks like `(None, None) => None` catches _any_ subtree in the segment in a similar way, not just the overall root? 
If both child nodes are `None` we do not always want to use `None` as the parent hash.

Does this happen somewhere else in the code?

Reply:
> Oh I get it now. We defer this to the next iteration.

Yes, correct.